### PR TITLE
Document data import and processing

### PR DESCRIPTION
**Pull request by flexik**
This PR documents data import and processing in Sirius.

This is a first draft and definitely not a final version.

### CODE REVIEW

**Comment by flexik** [comment]
@jirutka please have a look

File: docs/data-import-and-processing.adoc [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+Then a matching event is looked up in the database with the tuple `(source_entity_id, sequence_id)` instead of just with `source_entity_id`.
+After that a check is performed for extra events with sequence_id higher than highest generated sequence_id in group.
+These events are then marked as deleted, because they were not generated in the current dataset.
+
+=== Deleted events
+
+The case described above handles removal of extra events, that are no longer generated, because some event or planning parameters changed.
+But what about entities that were completely removed in KOS? They do not appear in import, thus there is no matching performed at all, so the approach described above doesn't help.
+When events are synchronized, all event ids that were processed (created, updated or unchanged) during current import run are temporarily stored.
+After event synchronization is complete, you the database is queried for events with ids that were not included the processed set.
```
Comment (jirutka) [inline docs/data-import-and-processing.adoc]: Typo: ~~you~~ the database is queried…

Comment (jirutka) [inline docs/data-import-and-processing.adoc]: …were not included **in** the processed set.

File: docs/data-import-and-processing.adoc [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+Returned events are no longer generated and are marked as deleted, because their source entity was removed.
+
+== Multi-faculty support
+
+Sirius can import and store timetable data for multiple faculties.
+Currently an import is performed for FIT and FEL, but other faculties can be easily added by configuring access to another instance of KOSapi.
```
Comment (jirutka) [inline docs/data-import-and-processing.adoc]: I think that there should be _**the**_ import.

**Comment by jirutka** [comment]
Great work @flexik! 😺

You said that it’s not a final version; what else are you going to add?

**Comment by jirutka** [comment]
@jnv Could you please finish the second part of the review?

**Comment by jnv** [comment]
Done.

File: docs/data-import-and-processing.adoc [hunk 3]
```diff
@@ -62,31 +62,30 @@ This means that you can have `update_other` turned on and `update_parallels` dis
...
 
-The external data can be changed or deleted at any time, so the synchronization process has to account for that.
-KOS(api) does not provide any diffs or changesets, so full synchronization has to be performed every time an update is required.
-Synchronization is usually scheduled multiple times a day, so that the most recent data are provided.
+The external data can be changed or deleted any time, so the synchronization process has to account for that.
+KOS(api) does not provide any diffs or changesets, so we need to performa a full synchronization every time an update is required.
+Synchronization happens multiple times a day to provide up-to-date data.
 
-When synchronizing with existing database data, events are generated as usual and when saving them, an attempt is made to match them with the existing events in the database.
+We synchronize existing database data as follows: events are generated as usual. When the generated events are saved to dabase, an attempt is made to match them with the existing events in the database.
```
Comment (jirutka) [inline docs/data-import-and-processing.adoc]: This sentence is kinda weird: “We synchronize existing database data **as follows: events are generated as usual.**”

Typo: …are saved to **dabase**…

Comment (jnv) [inline docs/data-import-and-processing.adoc]: Hm, yeah, doesn't make much sense. I think we could just completely replace this part like this:

> The goal of synchronization is to match and update existing events in database with generated events.

File: docs/data-import-and-processing.adoc [hunk 4]
```diff
@@ -63,15 +67,17 @@ This means that you can have `update_other` turned on and `update_parallels` dis
...
 
 The external data can be changed or deleted any time, so the synchronization process has to account for that.
-KOS(api) does not provide any diffs or changesets, so we need to performa a full synchronization every time an update is required.
+KOS(api) does not provide any diffs or changesets, so we need to perform a full synchronization every time an update is required.
 Synchronization happens multiple times a day to provide up-to-date data.
 
-We synchronize existing database data as follows: events are generated as usual. When the generated events are saved to dabase, an attempt is made to match them with the existing events in the database.
+The synchronization consists of two main phases.
+In the first phase, Sirius fetches source data from KOSapi and generates _events_.
+Then it tries to match the generated events with the existing ones in the database.
```
Comment (jirutka) [inline docs/data-import-and-processing.adoc]: @jnv What do you think?

File: docs/data-import-and-processing.adoc (hunk 3 above)
Comment (flexik) [inline docs/data-import-and-processing.adoc]: I aggree. The process is explained in detail in the following paragraph anyway. Maybe change the sentence a bit:
> The goal of synchronization is to match and update existing events in database with __newly__ generated events.

File: docs/data-import-and-processing.adoc [hunk 5]
```diff
@@ -62,31 +62,30 @@ This means that you can have `update_other` turned on and `update_parallels` dis
...
 If a match is found and the data is the same, there is no need to update anything.
-In a case there is a match and the event data differs, existing event data in the database had to be updated.
-And in a case there is no match at all, a new event with generated data is created.
+If there is a match and the event data differs, existing event in the database is updated.
+In case there is no match at all, a new event is created.
 
-Event matching is designed so that it provides the most stable matching possible (meaning it does not generate new events when not necessary).
-For data sources which have one-to-one mapping with resulting events (i.e. exams, course events) this is pretty easy.
-The source entity ID (for example an exam ID) is saved in the generated event and in the consecutive runs a match is performed against it.
+Event matching strives to be stable, i.e. it aims to update existing events instead of erasing and generating new events.
```
Comment (flexik) [inline docs/data-import-and-processing.adoc]: Maybe replace `erasing and generating` with `erasing and creating` here. Events are generated always during the import and then they are either saved as new records or update existing ones.

File: docs/data-import-and-processing.adoc (hunk 3 above)
Comment (flexik) [inline docs/data-import-and-processing.adoc]: Never mind, I've updated the current version of this paragraph a little.

**Pull request merged**